File: frame/src/main/java/com/dreampany/frame/data/misc/StateMapper.java
package com.dreampany.frame.data.misc;

import com.dreampany.frame.data.model.State;
import com.dreampany.frame.misc.SmartCache;
import com.dreampany.frame.misc.SmartMap;
import com.dreampany.frame.misc.StateAnnote;

import javax.inject.Inject;

/**
 * Created by Hawladar Roman on 8/9/2018.
 * BJIT Group
 * hawladar.roman@bjitgroup.com
 */
public class StateMapper {

    //keeps only LRU
    private final SmartMap<String, State> map;
    private final SmartCache<String, State> cache;

    @Inject
    StateMapper(@StateAnnote SmartMap<String, State> map,
                @StateAnnote SmartCache<String, State> cache) {
        this.map = map;
        this.cache = cache;
    }

    public boolean isExists(String id) {
        return map.contains(id);
    }

    public boolean isExists(String id, String type, String subtype, String state) {
        if (isExists(id)) {
            State item = getItem(id);
            return item.hasProperty(type, subtype, state);
        }
        return false;
    }

    public void putItem(State item) {
        map.put(item.getId(), item);
    }

    public State getItem(String id) {
        return map.get(id);
    }

    public State getItem(String id, String type, String subtype, String state) {
        if (isExists(id, type, subtype, state)) {
            return getItem(id);
        }
        return null;
    }
}

File: frame/src/main/kotlin/com/dreampany/frame/data/misc/StateMapper.kt
package com.dreampany.frame.data.misc

import com.dreampany.frame.data.model.State
import com.dreampany.frame.misc.SmartCache
import com.dreampany.frame.misc.SmartMap
import com.dreampany.frame.misc.StoreAnnote
import javax.inject.Inject
import javax.inject.Singleton

/**
 * Created by roman on 2019-07-25
 * Copyright (c) 2019 bjit. All rights reserved.
 * hawladar.roman@bjitgroup.com
 * Last modified $file.lastModified
 */
@Singleton
class StateMapper @Inject constructor(
    @StoreAnnote val map: SmartMap<String, State>,
    @StoreAnnote val cache: SmartCache<String, State>
) {

    fun isExists(id: String): Boolean {
        return map.contains(id)
    }

    fun isExists(id: String, type: String, subtype: String, state: String): Boolean {
        if (isExists(id)) {
            val item = getItem(id)
            return item.hasProperty(type, subtype, state)
        }
        return false
    }

    fun putItem(item: State) {
        map.put(item.id, item)
    }

    fun getItem(id: String): State {
        return map.get(id)
    }

    fun getItem(id: String, type: String, subtype: String, state: String): State? {
        return if (isExists(id, type, subtype, state)) {
            getItem(id)
        } else null
    }
}